README.md
# JobPostFeeder

There are two scripts: demo.py and jobs.py  

## demo.py

Retrieve job postings from Dice.com and show how many job postings for each cities/town in a browser.  

#### How to run
On Linux, go to the jobs directory and run `./demo.py`  
  
On Windows,  
``>python ./demo.py``
  
If it complains that you don't have BeautifulSoup and/or Flask, run
``pip install -r requirements.txt``
  
  
## jobs.py
  
Retrieve job postings from Dice.com and show as JSON objects.  

You can search for job postings with the keys like:  
* skill
* zip code
* key words in job description
* age of posting in days
  
The information about job postings such as company names and titles are returned in the JSON format. Each line has a single JSON object. 

#### How to run
On Linux, go to the jobs directory and run `./jobs.py`  
ex)  
``$ ./jobs.py -k "big data" -z 02140 -a 7 -S age -D -n``
  
On Windows,  
``>python jobs.py -k "big data" -z 02140 -a 7 -S age -D -n``
  
If it complains that you don't have BeautifulSoup and/or Flask, run
``pip install -r requirements.txt``
  
#### Command line options
* `-h, --help`  
   show the help messagne and quit  
* `-k KEYWORD [KEYWORD ...], --keyword KEYWORD [KEYWORD ...]`  
  specify search keywords for job description  
* `-s SKILL [SKILL ...], -skill SKILL [SKILL ...]`  
  specify search text for the jobs skill property  
* `-z ZIPCODE, --zipcode ZIPCODE`  
  specify a zipcode as the center of 40 mile radius  
* `-a AGE, -- age AGE`  
  specify a posting age (a.k.a. days back)  
* `-S {age,jobtitle,company,location}, --sort {age,jobtitle company,location}`  
  specify a sort key  
* `-D, --descending`  
  sort in descending order; without this, sort in scending order  
* `-n, --no_description`  
  don't get detail job description; return the URL instead  
* `-d, --direct`  
  return direct hire jobs only.  
  
#### Returned data
<table>
<tr><th>property</th><th>description</th></tr>
<tr><td>company</td><td>company or recuriter's name</td></tr>
<tr><td>postdate</td><td>the date of post</td></tr>
<tr><td>title</td><td>job title</td></tr>
<tr><td>location</td><td>location of the position</td></tr>
<tr><td>description</td><td>the detailed description about the position</td></tr>
<tr><td>posturl</td><td>the url to the job posting</td></tr>
</table>

#### Reference
* Dice jobsearch API  
  http://www.dice.com/common/content/util/apidoc/jobsearch.html


changes.md
#  Changes to migrate JobPostFeeder to Python 3.x


## jobs.py
## jobfeeder/jobfeed.py
## webdemo/handler.py
## webdemo/model.py
* **fix print commands -- print is a function in Python 3.x**


## jobfeeder/dice.py
* **fix print commands -- print is a function in Python 3.x**
* **fix urllib usage -- now urllib package is split into urllib.request, urllib.parse, and urllib.error**
* **fix use of basestring -- this type does not exist in Python 3.x**
* **fix definition of iterator -- now use \_\_iter\_\_() not iter()**
* **fix loading of JSON response**


## webdemo/app.py
* **fix the import of Flask extensions -- new package structure**


## webdemo/democonfig.py
* **add SQLALCHEMY_TRACK_MODIFICATIONS definition to suppress annoying warning**
* **print a couple of logging lines (commented out)**


## changes.md (this file)
* **list of changes made to migrate to Python 3.x**


## .gitignore
* **ignore webdemo/migrations folder -- not sure it is ever used**



jobs/jobs.py
#!/usr/bin/env python

'''Search for job postings

@author Hideki Ikeda
@created Apr 19, 2015

Command line options:
    -h, --help        show the help messagne and quit
    -k KEYWORD [KEYWORD ...], --keyword KEYWORD [KEYWORD ...]
                      specify search keywords for job description
    -s SKILL [SKILL ...], -skill SKILL [SKILL ...]
                      specify search text for the jobs skill property
    -z ZIPCODE, --zipcode ZIPCODE
                      specify a zipcode as the center of 40 mile radius
    -a AGE, -- age AGE
                      specify a posting age (a.k.a. days back)
    -S {age,jobtitle,company,location}, --sort {age,jobtitle company,location}
                      specify a sort key
    -D, --descending  sort in descending order; without this, sort in scending
                      order
    -n, --no_description
                      don't get detail job description; return the URL instead
    -d, --direct      return direct hire jobs only.
'''

import argparse

from jobfeeder import JobFeeder
from jobfeeder import idx_company
from jobfeeder import idx_postdate
from jobfeeder import idx_title
from jobfeeder import idx_location
from jobfeeder import idx_description


if __name__=='__main__':
    parser = argparse.ArgumentParser(
            description='Search for job posting and show them in the JSON format',
            epilog='ex) jobs.py -s c++ java -k "big data" -z 02140 -a 14 -S age -D')
    parser.add_argument('-k', '--keyword', nargs='+',
            help='specify search keywords for job description')
    parser.add_argument('-s', '--skill', nargs='+',
            help='specify search text for the jobs skill property')
    parser.add_argument('-z', '--zipcode', 
            help='specify a zipcode as the center of 40 mile redius')
    parser.add_argument('-a', '--age', type=int,
            help='specify a posting age (a.k.a. days back)')
    parser.add_argument('-S', '--sort',
            choices=['age', 'jobtitle', 'company', 'location'],
            help='specify a sort key')
    parser.add_argument('-D', '--descending', action='store_true',
            help='sort in descending order; without this, sort in ascending order')
    parser.add_argument('-n', '--no_description', action='store_true',
            help="don't get detailed job description; return the URL instead")
    parser.add_argument('-d', '--direct', action='store_true',
            help='return direct hire jobs only.')

    args = parser.parse_args()
    feeder = JobFeeder(direct=args.direct, skill=args.skill,
                       zipcode=args.zipcode, text=args.keyword,
                       age=args.age, sort=args.sort, sd=args.descending,
                       description=not args.no_description)

    for post in feeder:
        print(post)
